projects/novo-elements/src/elements/tabbed-group-picker/TabbedGroupPicker.ts

```diff
@@ -0,0 +1,345 @@
+import { Component, EventEmitter, Input, OnInit, Output, ChangeDetectionStrategy, ChangeDetectorRef, AfterViewInit } from '@angular/core';
+import { BehaviorSubject } from 'rxjs';
+import { debounceTime } from 'rxjs/operators';
+import { Helpers, binarySearch } from '../../utils/Helpers';
+import { NovoLabelService } from '../../services/novo-label-service';
+import { ScrollDispatcher, CdkScrollable, ExtendedScrollToOptions } from '@angular/cdk/scrolling';
+
+export type TabbedGroupPickerSchema = {
+  typeName: string;
+  typeLabel: string;
+  valueField: string;
+  labelField: string;
+  scrollOffset?: number;
+} & (ParentSchema | ChildSchema);
+
+export type ParentSchema = {
+  childTypeName: string;
+  data: Array<ParentOption>;
+};
+
+type ParentOption = {
+  selected?: boolean;
+  indeterminate?: boolean;
+  children: Array<{ selected?: boolean }>;
+} & { [key: string]: any };
+
+export type ChildSchema = {
+  data: Array<{ selected?: boolean } & { [key: string]: any }>;
+};
+
+export type TabbedGroupPickerQuickSelect = {
+  label: string;
+  selected?: boolean;
+  childTypeName?: string;
+  children?: (({ selected?: boolean } & { [key: string]: any }) | (number))[];
+  all?: boolean;
+};
+
+export type QuickSelectConfig = { label: string; items: TabbedGroupPickerQuickSelect[] };
+
+@Component({
+  selector: 'novo-tabbed-group-picker',
+  templateUrl: './TabbedGroupPicker.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class NovoTabbedGroupPickerElement implements OnInit, AfterViewInit {
+  @Input() buttonConfig: {
+    theme: string;
+    side: string;
+    icon: string;
+    label: string;
+  };
+  @Input() schemata: TabbedGroupPickerSchema[];
+  @Input() quickSelectConfig: QuickSelectConfig;
+
+  @Input() title: string = 'tabbed-group-picker'; // need unique information in order to find scroll container from scroll dispatcher
+  displaySchemata: TabbedGroupPickerSchema[];
+
+  @Output() selectionChange: EventEmitter<any> = new EventEmitter<any>();
+
+  displaySchemaIndex: number = 0;
+  scrollableInstance: CdkScrollable;
+
+  filterText: BehaviorSubject<string> = new BehaviorSubject('');
+
+  loading = true;
+
+  showClearAll: boolean = false;
+
+  get displaySchema(): TabbedGroupPickerSchema {
+    return this.displaySchemata[this.displaySchemaIndex];
+  }
+  set displaySchema(schema: TabbedGroupPickerSchema) {
+    this.displaySchemaIndex = this.schemata.map(({ typeName }) => typeName).indexOf(schema.typeName);
+  }
+
+  get viewportClass(): string {
+    return `virtual-scroll-viewport ${this.title}`;
+  }
+
+  get virtualScrollItemSize() {
+    const em = 3.1; // this corresponds to TabbedGroupPicker.scss .tabbed-group-picker-column-container.tabbed-group-picker-column &.right novo-list-item.height
+    const defaultFontSize = 16; // fingers crossed
```

Comment:
I don't like comments like this in the code base.

Reply:
Yup, I'll clean it up.